### ISSUE TITLE
Use gpkg.in for versioning?

### ISSUE THREAD

**Issue by amoghe** [jinzhu/gorm#373]
[This](http://labix.org/gopkg.in) helps with API versioning. The idea is that the package maintainers release new versions of the pkg/lib/api and folks downstream (like me) can consume new versions by bumping up the version of the gpkg.in url.

A good example of a package using this:  https://gopkg.in/alecthomas/kingpin.v1

**Comment by jinzhu** [comment]
Hello @amoghe 

Although the idea behind gopkg is interesting, it can't be used for private repos, so I don't think it could be used as seriously package manager.

Why not using other tools like `gom` https://github.com/mattn/gom, it could works with git commit hash and don't rely other third party services (more risky)

Even you don't need to deal with private packages, it still not good to use gopkg as the package manager. For example:

If package owner using branch as version, and he push a commit to that branch hope to fix something, but it will break your system, AFAIK, you can't ignore the commit with gopkg, yes? maybe you don't even know the bug because the commit is not included in your development machine but only included on the server.

**Comment by amoghe** [comment]
@jinzhu I'm not sure I follow why its not a good option. For example, several golang projects are using it quite successfully to provide API stability guarantees. I agree that it is not useful for private repo's , but gorm is not private, right?

> If package owner using branch as version, and he push a commit to that branch hope to fix something, but it will break your system

Yes, it will break, but the point is not to avoid breakage. gokpkg.in wants to give you API stability. The point of software versioning is to give you guarantees that the API (or whatever interface the software offers  - e.g. - http API's , or function calls, or methods) is stable and will not change. So for gorm, this would have been useful for me when you added a new method on the `Model` , but I was using an older version of gorm, so that method was not available to me, and it was not clear why the compile was failing when i tried to use it. Later I had to go and check the git history to realize that I was using an older version. If gorm was using a version number , then it would have been easy to check that I need to update to a newer version.

So in short, the point of gopkg.in is to provide API versioning. New versions can have bugs and new software can break, but that is outside the purview of version management.

**Comment by jaytaylor** [comment]
You should probably use godep if you want to pin package versioning rather than asking package developers and maintainers to change their workflow.

I want to see gorm keep getting more and more awesome!  Not that worried about breakage, jinzhu has demonstrated very sensible and reasonable judgement imo.

What is the concern, problem or pay you've been caused by the current strategy?

**Comment by amoghe** [comment]
> I want to see gorm keep getting more and more awesome! 

Absolutely, I think we're all in agreement here.

> What is the concern, problem or pain that's been caused by the current strategy?

I don't have a "current" strategy , but I'm trying to see which one would work. Looks like there are 2 distinct camps.

1. Use gopkg.in. This approach doesn't aim for reproducible builds. The idea here is that you don't pin yourself to a particular version of the code, instead you depend on a certain version of the API. So for example, I want to use (some imaginary pkg) ```libfoo``, as a developer I care about using the most recent/stable version, say `v2`, but I don't necessarily care about the exact git hashref. This way my application picks up new versions of the library (```2.1```, ```2.2```, ```2.x```) every time it is built, but the build doesn't break because the API is stable. This is how the version number provides *API stability*.

2. Use godep. This approach you pin down a particular version of the source code you want to build against. This approach gives you reproducible builds, obviously. However, this also means that in order to move to a new version the consumer of ```libfoo``` needs to move to a new git hashref (which may be more stable, and have bugfixes), but you also run the risk of bumping up to a new version of the API (and thus have to rewrite a bunch of code).

> You should probably use godep if you want to pin package versioning  

So it seems like these two approaches are significantly different and godep is not a drop-in replacement for gopkg.in (even though at first glance, it seems like they're achieving the same thing). The author of gopkg.in even has examples of how these two tools can work in conjunction with each other. My original question was to determine if the second approach (using gopkg.in) was something @jinzhu would consider. APIs almost always change (no API is perfectly future-proof) , and what I'm looking for in the short term is API stability. 

> rather than asking the developers and maintainers to change their workflow and existing system

I hope we're on the same page now. If you still think this is an unreasonable request I'd like to hear your (and @jinzhu 's) thoughts on that. Maybe there is something I've missed.

Additional thoughts:

Of course, it adds the additional burden on the maintainer to have branches in the source tree and fix bugs in more than one branches. However other open source developers have mitigated this problem by maintaining at most one version that is behind the current latest (deprecating older versions aggressively). The golang community moves fast, and folks would move to a new version in reasonable time, so this might not be a bad thing.
